pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,169 @@
+
+    private long waitTimeMs;
+    private volatile boolean isRunning = true;
+    private NiFiConfig niFiConfig;
+    private SiteToSiteClientConfig clientConfig;
+
+    Thread runnerThread;
+
+    @Override
+    public void open(Map<String, Object> config, SourceContext sourceContext) throws Exception {
+        niFiConfig = NiFiConfig.load(config);
+        if (niFiConfig.getUrl() == null
+                || niFiConfig.getPortName() == null) {
+            throw new IllegalArgumentException("Required property not set.");
+        }
+
+        waitTimeMs = niFiConfig.getWaitTimeMs();
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+
+        this.start();
+    }
+
+    @Override
+    public void close() throws Exception {
+        this.isRunning = false;
+        if (runnerThread != null) {
+            runnerThread.interrupt();
+            runnerThread.join();
+            runnerThread = null;
+        }
+    }
+
+    protected void start() throws Exception {
+        runnerThread = new Thread(new ReceiveRunnable());
+        runnerThread.setName("Nifi Source Thread");
+        runnerThread.start();
+    }
+
+
+    class ReceiveRunnable implements Runnable {
+
+        public ReceiveRunnable() {
+        }
+
+        @Override
+        public void run() {
+            try {
+                final SiteToSiteClient client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+                try {
+                    while (isRunning) {
+                        final Transaction transaction = client.createTransaction(TransferDirection.RECEIVE);
+                        DataPacket dataPacket = transaction.receive();
+                        if (dataPacket == null) {
+                            transaction.confirm();
+                            transaction.complete();
+
+                            // no data available. Wait a bit and try again
+                            try {
+                                Thread.sleep(waitTimeMs);
+                            } catch (InterruptedException e) {
+                            }
+                            continue;
+                        }
+
+                        final List<NiFiDataPacket> dataPackets = new ArrayList<>();
+                        do {
+                            // Read the data into a byte array and wrap it along with the attributes
+                            // into a NiFiDataPacket.
+                            final InputStream inStream = dataPacket.getData();
+                            final byte[] data = new byte[(int) dataPacket.getSize()];
+                            StreamUtils.fillBuffer(inStream, data);
+
+                            final Map<String, String> attributes = dataPacket.getAttributes();
```

Comment:
It's not clear to me how/if we are getting the attributes from NiFi. Can you add a test case that verifies we are getting the attributes?

Reply:
TestSinkAndSource
``` java
       Record<NiFiDataPacket> readRecord = source.read();
        NiFiDataPacket recordValue = readRecord.getValue();
        System.out.println(recordValue.getAttributes());
        System.out.println(new String(recordValue.getContent()));
```
sys out
```java
{path=./, filename=cb13f0b1-b370-4be7-8196-6e444a706858, uuid=cb13f0b1-b370-4be7-8196-6e444a706858, s2s.address=localhost:53509, s2s.host=localhost}
Data from Pulsar
```
next commit I can Assert.assertNotNull(recordValue.getAttributes());